src/Runner.Worker/ExecutionContext.cs

```diff
@@ -411,21 +432,32 @@ public TaskResult Complete(TaskResult? result = null, string currentOperation =
                 this.Warning($"The job has experienced {TimeSpan.FromMilliseconds(_totalThrottlingDelayInMilliseconds).TotalSeconds} seconds total delay caused by server throttling.");
             }
 
+            DateTime now = DateTime.UtcNow;
             _record.CurrentOperation = currentOperation ?? _record.CurrentOperation;
             _record.ResultCode = resultCode ?? _record.ResultCode;
-            _record.FinishTime = DateTime.UtcNow;
+            _record.FinishTime = now;
             _record.PercentComplete = 100;
             _record.Result = _record.Result ?? TaskResult.Succeeded;
             _record.State = TimelineRecordState.Completed;
 
+            // Before our main timeline's final QueueTimelineRecordUpdate,
+            //    inject any issues collected by embedded ExecutionContexts.
+            if (!this.IsEmbedded)
+            {
+                foreach (var issue in _embeddedIssueCollector)
+                {
+                    AddIssue(issue);
```

Comment:
The AddIssue function has a side effect of writing a line to the job logs. Does this change make those changes write twice?

As a part of this refactor, we may want to remove the side effect from the AddIssue function, have one function that
- masks any secrets
- writes to a log
- returns a hydrated timelinerecordissue object
And a second that
- Adds that timelinerecordissue to the timeline record and queues an update.

Reply:
It doesn't appear that anything will get written to the job log as a result of this call.  All the write logic is gated on a non-null `logMessage` being passed as the second (optional) parameter.

see https://github.com/actions/runner/blob/7036627d4717dadc479a2917b6f79859681fd606/src/Runner.Worker/ExecutionContext.cs#L581

Reply:
I created a one-off exploration of something along those lines [here](https://github.com/actions/runner/pull/2378)

Reply:
This should be fixed, now.  I decided not to tease apart the logging calls from `AddIssue` since all callers except one desire/expect that behavior.  Instead, I replaced the log message override string parameter with a small struct that allows the caller to clearly communicate their intent.